### PR TITLE
Small fixes to database pruning and updating

### PR DESCRIPTION
**Pull request by nickjcroucher**
- Fix to `--remove-samples`, which did not work due to missing parentheses
- Updated QC mode to also prune samples out of the graph with `prune_graph` - e.g. for trimming a database, or for removing inappropriate samples only identified after visualisation (I assume networks code will be optimised in the future, so this is a little rough and ready)
- Updated `unwords.py` to no longer use sets - https://stackoverflow.com/questions/15837729/random-choice-from-set

### CODE REVIEW

**Comment by johnlees** [comment]
I'll have a look at this next week. We also need to co-ordinate with #322 which I had almost finished to fix the latter issue too, maybe I can just remove it from there.

Also, was #304 supposed to have been merged at some point, or does this replace that?

**Comment by nickjcroucher** [comment]
Apologies, hadn't seen #322 - I don't get notifications of issues as I used to for some reason. These were just the changes I needed to make to prepare data for @absternator. #304 is more a rumination on the nature of time and mortality at this point, but I'll pull across the changes from `master` into it when I get a chance.

File: PopPUNK/bgmm.py [hunk 1]
```diff
@@ -45,10 +45,36 @@ def fit2dMultiGaussian(X, dpgmm_max_K = 2):
...
+    assigned to it
+
+    Args:
+        means (numpy.array)
+            K x 2 array of mixture component means
+        assignments (numpy.array)
+            Sample cluster assignments
+        rank (int)
+            Which label to find, ordered by distance from origin. 0-indexed.
+            (default = 0)
```
Comment (johnlees) [inline PopPUNK/bgmm.py:61]: I think ordered by number of assigned samples, not distance from the origin?

File: PopPUNK/models.py [hunk 2]
```diff
@@ -333,7 +334,7 @@ def fit(self, X, max_components):
         else:
             y = self.assign(self.subsampled_X, max_batch_size = self.max_batch_size)
         self.within_label = findWithinLabel(self.means, y)
-        self.between_label = findWithinLabel(self.means, y, 1)
+        self.between_label = findBetweenLabel_bgmm(self.means, y)
```
Comment (johnlees) [inline PopPUNK/models.py:337]: Why this change in behaviour? This would change some of the examples I think

File: PopPUNK/network.py [hunk 3]
```diff
@@ -1837,3 +1888,68 @@ def sparse_mat_to_network(sparse_mat, rlist, use_gpu = False):
...
+        graph_suffix = '.gt'
+
+    network_found = False
+    for graph_name in ['_core.refs_graph','_core_graph','_accessory.refs_graph','_accessory_graph','.refs_graph','_graph']:
+      network_fn = f"{prefix}/{os.path.basename(prefix)}" + graph_name + graph_suffix
+      if os.path.exists(network_fn):
+          network_found = True
+          sys.stderr.write("Loading network from " + network_fn + "\n")
+          passed_set = frozenset(passed)
+          G = load_network_file(network_fn, use_gpu = use_gpu)
```
Comment (johnlees) [inline PopPUNK/network.py:1928]: Surely we already have a function to open a network somewhere? I think it would be better for this function to take a network rather than a prefix as input

File: PopPUNK/network.py [hunk 4]
```diff
@@ -1837,3 +1888,68 @@ def sparse_mat_to_network(sparse_mat, rlist, use_gpu = False):
...
+
+
+def prune_graph(prefix, reflist, passed, output_db_name, threads, use_gpu):
+    """Keep only the specified sequences in a graph
+
+    Args:
+       prefix (str)
+           Name of directory containing network
+       reflist (list)
+           Ordered list of sequences of database
```
Comment (johnlees) [inline PopPUNK/network.py:1900]: I don't like that this has to be ordered, or I don't understand it. Why is this needed? Seems easy to get wrong

**Review by johnlees — CHANGES_REQUESTED** [review]

File: PopPUNK/assign.py [hunk 5]
```diff
@@ -753,7 +753,7 @@ def assign_query_hdf5(dbFuncs,
             storePickle(combined_seq, combined_seq, True, None, dists_out)
 
             # Clique pruning
-            if model.type != 'lineage':
+            if model.type != 'lineage' and os.path.isfile(ref_file_name):
```
[the remark this answers is not in the document]
Comment (nickjcroucher) [reply PopPUNK/assign.py:756]: I don't think it changes the behaviour if there is no reference file, as in the current GPS database - I suppose it just depends what we want the behaviour to be in such a situation.

File: PopPUNK/__main__.py [hunk 6]
```diff
@@ -156,6 +156,9 @@ def get_options():
     refinementGroup.add_argument('--score-idx',
             help='Index of score to use [default = 0]',
             type=int, default = 0, choices=[0, 1, 2])
+    refinementGroup.add_argument('--summary-sample',
+            help='Number of sequences used to estimate graph properties [default = all]',
+            type=int, default = None)
```
[the remark this answers is not in the document]
Comment (nickjcroucher) [reply PopPUNK/__main__.py:161]: Good point, done in 962b198.

File: PopPUNK/models.py (hunk 2 above)
Comment (nickjcroucher) [reply PopPUNK/models.py:337]: This harmonises the behaviour of BGMM and DBSCAN - identifying the between-strain cluster as that containing the largest number of points. Name is a bit ugly.

File: PopPUNK/bgmm.py (hunk 1 above)
Comment (nickjcroucher) [reply PopPUNK/bgmm.py]: Pointless. Removed in 6bbfdb7.

File: PopPUNK/network.py (hunk 3 above)
Comment (nickjcroucher) [reply PopPUNK/network.py:1928]: Agree that network interface could do with a refresh. Can we set up a new issue/branch for this (and include the GPU fast prune update from #322?

File: PopPUNK/network.py [hunk 7]
```diff
@@ -1837,3 +1888,68 @@ def sparse_mat_to_network(sparse_mat, rlist, use_gpu = False):
...
+                      reference_vertex[vertex] = True
+                  else:
+                      reference_vertex[vertex] = False
+              G_new = gt.GraphView(G, vfilt = reference_vertex)
+              G_new = gt.Graph(G_new, prune = True)
+          save_network(G_new,
+                      prefix = output_db_name,
+                      suffix = '_graph',
+                      use_graphml = False,
+                      use_gpu = use_gpu)
```
[the remark this answers is not in the document]
Comment (nickjcroucher) [reply PopPUNK/network.py:1953]: As above.

File: PopPUNK/visualise.py [hunk 8]
```diff
@@ -137,6 +137,7 @@ def get_options():
     other.add_argument('--gpu-dist', default=False, action='store_true', help='Use a GPU when calculating distances [default = False]')
     other.add_argument('--gpu-graph', default=False, action='store_true', help='Use a GPU when calculating graphs [default = False]')
     other.add_argument('--deviceid', default=0, type=int, help='CUDA device ID, if using GPU [default = 0]')
+    other.add_argument('--tmp', default=None, type=str, help='Directory for large temporary files')
```
[the remark this answers is not in the document]
Comment (nickjcroucher) [reply PopPUNK/visualise.py]: Good point, done in 77537af.

File: PopPUNK/utils.py [hunk 9]
```diff
@@ -542,12 +542,19 @@ def decisionBoundary(intercept, gradient):
             which intercepts the boundary
         gradient (float)
             Gradient of the line
+        adj (float)
+            Distance by which to shift the interception point
```
[the remark this answers is not in the document]
Comment (nickjcroucher) [reply PopPUNK/utils.py:546]: Used in https://github.com/bacpop/PopPUNK/blob/77537af9479716412c4b625dea388a43a0d550d1/PopPUNK/refine.py#L117 to harmonise behaviour of `unconstrained` with other refinement modes.

[the remark this answers is not in the document]
Comment (nickjcroucher) [reply PopPUNK/utils.py]: Better definition in 2054e6c.

File: PopPUNK/network.py [hunk 10]
```diff
@@ -1837,3 +1888,68 @@ def sparse_mat_to_network(sparse_mat, rlist, use_gpu = False):
...
+       threads (int)
+            Number of CPU threads to use when recalculating random match chances
+            [default = 1].
+       use_gpu (bool)
+            Whether graph is a cugraph or not
+            [default = False]
+
+    Returns:
+       vlist (list)
+           List of integers corresponding to nodes
```
[the remark this answers is not in the document]
Comment (nickjcroucher) [reply PopPUNK/network.py]: Well spotted, removed in 4d8f1b2.

File: PopPUNK/__main__.py [hunk 11]
```diff
@@ -424,18 +429,18 @@ def main():
         if args.remove_samples:
             with open(args.remove_samples, 'r') as f:
                 for line in f:
-                    fail_unconditionally[line.rstrip] = ["removed"]
+                    sample_to_remove = line.rstrip()
+                    if sample_to_remove in refList:
+                        fail_unconditionally[sample_to_remove] = ["removed"]
 
         # assembly qc
-        sys.stderr.write("Running sequence QC\n")
```
[the remark this answers is not in the document]
Comment (nickjcroucher) [reply PopPUNK/__main__.py:430]: Yes, that's right.

File: PopPUNK/network.py (hunk 4 above)
Comment (nickjcroucher) [reply PopPUNK/network.py:1900]: Needs to be ordered because graph nodes are just integers. I did manually check this was working on a toy example as I was also nervous. This is only called within `remove_qc_fail`, which uses the ordered `names` for the distance matrix pruning, so hopefully limited scope for error beyond our current implementation.

**Comment by nickjcroucher** [comment]
Thank you for the comments @johnlees! I think I've addressed most of them - would it now make sense:

- merge in this and #322
- set up an issue for updating the network functions, as suggested here, and to enable GPU-based fast pruning (yes, the can would be kicked dow the road)
- set up a new branch to test whether we can get beebop running on the full database using the modifications you suggested, so we can keep @absternator busy?

File: PopPUNK/network.py (hunk 4 above)
Comment (johnlees) [reply PopPUNK/network.py:1900]: Can we at least add a check that it is ordered in the function?

**Comment by johnlees** [comment]
One final response in my comments, just on checking input is sorted

> Thank you for the comments @johnlees! I think I've addressed most of them - would it now make sense:
> 
>     * merge in this and [unword consonants as list #322](https://github.com/bacpop/PopPUNK/pull/322)

Let's merge this first. I still wanted to test a couple of things on that branch (should have time in a couple of weeks), then I can just resolve and merge problems on that one myself when it's ready.

>     * set up an issue for updating the network functions, as suggested here, and to enable GPU-based fast pruning (yes, the can would be kicked down the road)

Yep, good plan

>     * set up a new branch to test whether we can get beebop running on the full database using the modifications you suggested, so we can keep @absternator busy?

Also makes sense!

**Review by johnlees — APPROVED** [review]
